### ISSUE TITLE
Seems per_token_loss is always zero?

### ISSUE THREAD

**Issue by nkqiaolin** [dhcode-cpp/X-R1#11]
https://github.com/dhcode-cpp/X-R1/blob/7b2f2d5543f1ca1e88d263b8ba479d510778fadc/src/x_r1/x_grpo_trainer.py#L200C9-L200C64

In above code, per_token_loss seems always zero? Appreciate so much if you would provide more explanations about this line. Meanwhile, where is the (PI_old) in the original GRPO paper please?

Thanks!

**Comment by nkqiaolin** [comment]
I found the answer in TRL discussions:  https://github.com/huggingface/trl/issues/2608

**Comment by dhcode-cpp** [comment]
`per_token_logps.detach()` as `PI_old` is constant value, the ratio 

and `per_token_logps` as current new policy, the gradient will update.

**Comment by dhcode-cpp** [comment]
`torch.exp(per_token_logps - per_token_logps.detach())`  value always equal `1`, because `exp(0) = 1`

**Comment by nkqiaolin** [comment]
> `torch.exp(per_token_logps - per_token_logps.detach())` value always equal `1`, because `exp(0) = 1`

Yes，that's the key. Thanks.

**Comment by dhcode-cpp** [comment]
> > `torch.exp(per_token_logps - per_token_logps.detach())` value always equal `1`, because `exp(0) = 1`
> 
> Yes，that's the key. Thanks.

it's trl trick , the standard GRPO use clip ratio, maybe like PPO use minibatch to compute old-new policy ratio, but trl skip it.
